gammapy/stats/variability.py

```diff
@@ -281,3 +283,105 @@ def structure_function(flux, flux_err, time, tdelta_precision=5):
+    power_spectrum_params : dict , optional
+        Dictionary of parameters to be provided to the power spectrum function.
+
+    Returns
+    -------
+    time_series: `~numpy.ndarray`
+        Simulated time series.
+    time_axis: '~astropy.Quantity'
+        Time axis of the series in the same units as 'spacing'. It will be defined with length 'npoints', from 0 to
+        'npoints'*'spacing'.
+    Examples
+    --------
+    To pass the function to be used in the simlation one can use either the 'lambda' keyword or an extended definition.
+    Parameters of the function can be passed using the 'power_spectrum_params' keyword.
+    For example, these are three ways to pass a power law (red noise) with index 2:
+
+    >>> from gammapy.stats import TimmerKonig_lightcurve_simulator
+    >>> import astropy.units as u
+    >>> def powerlaw(x):
+    >>>     return x**(-2)
+    >>> def powerlaw_with_parameters(x, i):
+    >>>     return x**(-i)
+    >>> ts, ta = TimmerKonig_lightcurve_simulator(lambda x: x**(-2), 20, 1*u.h)
+    >>> ts2, ta2 = TimmerKonig_lightcurve_simulator(powerlaw, 20, 1*u.h)
+    >>> ts3, ta3 = TimmerKonig_lightcurve_simulator(powerlaw_withparameters, 20, 1*u.h, power_spectrum_params={"i":2})
+
+    References
+    ----------
+    ..[Timmer1995]"On generating power law noise", J. Timmer and M, Konig, section 3.
+    """
+    if not callable(power_spectrum):
+        raise ValueError(
+            "The power spectrum has to be provided as a callable function."
+        )
+
+    random_state = get_random_state(random_state)
+
+    frequencies = np.fft.fftfreq(npoints, spacing.value)
+
+    # To obtain real data only the positive or negative part of the frequency is necessary.
+    real_frequencies = np.sort(np.abs(frequencies[frequencies < 0]))
+
+    if power_spectrum_params:
+        periodogram = power_spectrum(real_frequencies, **power_spectrum_params)
+    else:
+        periodogram = power_spectrum(real_frequencies)
+
+    real_part = random_state.normal(0, 1, len(periodogram) - 1)
+    imaginary_part = random_state.normal(0, 1, len(periodogram) - 1)
+
+    # Nyquist frequency component handling
+    if npoints % 2 == 0:
+        idx0 = -2
+        random_factor = random_state.normal(0, 1)
+    else:
+        idx0 = -1
+        random_factor = random_state.normal(0, 1) + 1j * random_state.normal(0, 1)
+
+    fourier_coeffs = np.concatenate(
+        [
+            np.sqrt(0.5 * periodogram[:-1]) * (real_part + 1j * imaginary_part),
+            np.sqrt(0.5 * periodogram[-1:]) * random_factor,
+        ]
+    )
+    fourier_coeffs = np.concatenate(
+        [fourier_coeffs, np.conjugate(fourier_coeffs[idx0::-1])]
+    )
+
+    fourier_coeffs = np.insert(fourier_coeffs, 0, 0)
+    time_series = np.fft.ifft(fourier_coeffs).real
+
+    if time_series.max() < np.abs(time_series.min()):
+        time_series = -time_series
+
+    time_series = 0.5 + 0.5 * time_series / time_series.max()
```

Comment:
Is it not faster to make` 0.5 + 0.5 / time_series.max() * time_series` ? it avoids to access to the array twice...

Reply:
Not sure if it makes a significative difference but I added a commit to define tmax=time_series.max() so it is not computed twice.